### PR TITLE
fix(glean): always set page metrics + send ping when whoami returns or fails

### PR DESCRIPTION
**Pull request by caugner**
### Problem

1. The page metrics don't show up in the `action` event locally, if rumba is not running.
2. The page ping is never send locally, if rumba is not running.

Explanation: We don't set the page metrics (and send the `page` event, for that matter) unless we have `userData` (which requires rumba).

### Solution

1. Always set the page metrics immediately, without waiting for `userData`.
2. Send the `page` even as soon as we have `userData`, even if the "whoami" request failed.

---

### After

---

## How did you test this change?

See: https://debug-ping-preview.firebaseapp.com/pings/mdn-dev/7b8cda49-efd1-4828-9b1c-093be1e06b56

### CODE REVIEW

File: client/src/telemetry/glean-context.tsx [hunk 1]
```diff
@@ -132,15 +132,18 @@ export function useGleanPage() {
   const path = useRef<String | null>(null);
 
   return useEffect(() => {
-    if (!isServer && userData && path.current !== loc.pathname) {
+    if (!isServer) {
       path.current = loc.pathname;
```
Comment (caugner) [inline client/src/telemetry/glean-context.tsx:136]: Move down.

Comment (caugner) [reply client/src/telemetry/glean-context.tsx]: Resolved via af0d30e7f.

File: client/src/telemetry/glean-context.tsx [hunk 2]
```diff
@@ -132,15 +132,18 @@ export function useGleanPage() {
   const path = useRef<String | null>(null);
 
   return useEffect(() => {
-    if (!isServer && userData && path.current !== loc.pathname) {
-      path.current = loc.pathname;
-      gleanAnalytics.page({
+    if (!isServer) {
```
[the remark this answers is not in the document]
Comment (caugner) [reply client/src/telemetry/glean-context.tsx:135]: True

[the remark this answers is not in the document]
Comment (caugner) [reply client/src/telemetry/glean-context.tsx]: Resolved via f3f1edfe6d2486f9c9c48b97c18f273e42cb1074.

File: client/src/user-context.tsx [hunk 3]
```diff
@@ -236,10 +238,10 @@ export function UserDataProvider(props: { children: React.ReactNode }) {
     }
   }, [data]);
 
-  let userData = data || getSessionStorageData();
+  const userData = isLoading ? getSessionStorageData() : error ? null : data;
```
[the remark this answers is not in the document]
Comment (LeoMcA) [reply client/src/user-context.tsx:241]: Oh, fair enough, ignore this comment then

File: client/src/plus/offer-overview/offer-overview-subscribe/index.tsx [hunk 4]
```diff
@@ -238,15 +238,21 @@ function OfferDetails({
...
   return user.subscriptionType === subscriptionType;
 }
 
-function canUpgrade(user: UserData | null, subscriptionType: SubscriptionType) {
-  if (user === null || !user.isAuthenticated) {
+function canUpgrade(
+  user: UserData | null | undefined,
+  subscriptionType: SubscriptionType
+) {
+  if (!user || !user.isAuthenticated) {
```
Comment (caugner) [inline client/src/plus/offer-overview/offer-overview-subscribe/index.tsx:255]:
```suggestion
  if (!user?.isAuthenticated) {
```

File: client/src/plus/offer-overview/offer-overview-subscribe/index.tsx [hunk 5]
```diff
@@ -238,15 +238,21 @@ function OfferDetails({
   );
 }
 
-function isCurrent(user: UserData | null, subscriptionType: SubscriptionType) {
-  if (user === null || !user.isAuthenticated) {
+function isCurrent(
+  user: UserData | null | undefined,
+  subscriptionType: SubscriptionType
+) {
+  if (!user || !user.isAuthenticated) {
```
Comment (caugner) [inline client/src/plus/offer-overview/offer-overview-subscribe/index.tsx:245]:
```suggestion
  if (!user?.isAuthenticated) {
```

File: client/src/user-context.tsx (hunk 3 above)
Comment (LeoMcA) [inline client/src/user-context.tsx:241]: Brain's a bit slow today, perhaps this is easier to read?

```suggestion
  const userData = data || error ? null : getSessionStorageData();
```

**Review by LeoMcA — APPROVED** [review]

File: client/src/user-context.tsx (hunk 3 above)
Comment (caugner) [reply client/src/user-context.tsx:241]: I'm not sure, I tend to say it's more confusing. How about:

```suggestion
  let userData;
  if (isLoading) {
    userData = getSessionStorageData();
  } else if (error) {
    userData = null;
  } else {
    userData = data;
  }
```

Then again, I wonder if `userData` is `null` if `error` anyways?

Comment (LeoMcA) [reply client/src/user-context.tsx:241]: Nah, stick with the ternaries, let's keep `userData` a `const` :)

**Pull request merged**